lib: smf: make number of arguments constant for SMF_CREATE_STATE()

maass-hamburg opened the pull request:
Currently the number of arguments of `SMF_CREATE_STATE()` is dependent on the Kconfig options `CONFIG_SMF_ANCESTOR_SUPPORT` and `CONFIG_SMF_INITIAL_TRANSITION`, IMO this is bad, as it breaks existing code if a Kconfig option is enabled, that wasn't enabled when the existing code was written. 

Explicitly as the USB-C subsystem uses the SMF with `CONFIG_SMF_ANCESTOR_SUPPORT` enabled and `CONFIG_SMF_INITIAL_TRANSITION` disabled, so the user can’t use hierarchical state machine with support for initial transitions in the same project together with using the USB-C subsystem, as enabling `CONFIG_SMF_INITIAL_TRANSITION` would break the USB-C subsystem.

Therefore `SMF_CREATE_STATE()` should always have 5 arguments and the unused arguments should be set to `NULL`

Resolves: #71252

henrikbrixandersen reviewed (changes requested):
I think this is a good idea, but since it is a breaking API change, it needs to go through the process for introducing such changes. 

Also, this PR is missing an entry in the migration guide.

maass-hamburg committed:
lib: smf: constant number of arguments for SMF_CREATE_STATE()

This sets the number of arguments for SMF_CREATE_STATE() to always
be the same, independent of the selected Kconfig options.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
doc: releases: migration guide: 3.7: change of SMF_CREATE_STATE()

Mention change of SMF_CREATE_STATE().

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

henrikbrixandersen commented:
@maass-hamburg You need to follow the process described here: https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-breaking-api-changes

maass-hamburg committed:
doc: releases: release notes: 3.7: change of SMF_CREATE_STATE()

Mention change of SMF_CREATE_STATE().

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

jfischer-no commented:
> I think this is a good idea, but since it is a breaking API change, it needs to go through the process for introducing such changes.
> 
> Also, this PR is missing an entry in the migration guide.

Where has it been marked stable? https://docs.zephyrproject.org/latest/develop/api/overview.html

keith-zephyr commented on include/zephyr/smf.h:
```diff
@@ -55,25 +38,15 @@
- * @param _exit  State exit function
- */
-#define SMF_CREATE_STATE(_entry, _run, _exit) \
+#define SMF_CREATE_STATE(_entry, _run, _exit, _parent, _initial) \
```
This change should be squashed with the changes to the SMF_CREATE_STATE users.  Otherwise this breaks bisect-ability.

keith-zephyr commented on tests/lib/smf/src/test_lib_flat_smf.c:
```diff
@@ -213,10 +213,10 @@ static void state_d_exit(void *obj)
-	[STATE_B] = SMF_CREATE_STATE(state_b_entry, state_b_run, state_b_exit),
-	[STATE_C] = SMF_CREATE_STATE(state_c_entry, state_c_run, state_c_exit),
-	[STATE_D] = SMF_CREATE_STATE(state_d_entry, state_d_run, state_d_exit),
+	[STATE_A] = SMF_CREATE_STATE(state_a_entry, state_a_run, state_a_exit, NULL, NULL),
```
Squash this commit into the commit that modifies smf.h

keith-zephyr commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -1251,161 +1251,192 @@ static const struct smf_state pe_states[PE_STATE_COUNT] = {
 		NULL,
 		pe_sender_response_run,
 		pe_sender_response_exit,
+		NULL,
```
Squash into smf.h commit.

ycsin reviewed (commented):
Haven't looked at the code thoroughly, I wonder if we can use some macro `__VA_ARGS__` magic so that the API appears to be the same and we do not need to modify existing applications?

maass-hamburg commented:
> Haven't looked at the code thoroughly, I wonder if we can use some macro `__VA_ARGS__` magic so that the API appears to be the same and we do not need to modify existing applications?

The problem is, that `SMF_CREATE_STATE()` is a macro and not a function. Macros can't use `__VA_ARGS__` to get the single parameters.

glenn-andrews commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -248,6 +248,14 @@ Modem
+State Machine Framework
+=======================
+
+* The :c:macro:`SMF_CREATE_STATE` marcro now always takes 5 arguments. The amount of arguments is
```
macro not marcro

glenn-andrews commented on doc/releases/release-notes-3.7.rst:
```diff
@@ -229,6 +229,10 @@ Libraries / Subsystems
 
+* State Machine Framework
+
+  * The :c:macro:`SMF_CREATE_STATE` marcro now always takes 5 arguments.
```
macro not marcro

ycsin commented:
> The problem is, that `SMF_CREATE_STATE()` is a macro and not a function. Macros can't use `__VA_ARGS__` to get the single parameters.

i.e.:

```c
#ifdef CONFIG_SMF_ANCESTOR_SUPPORT
#ifndef CONFIG_SMF_INITIAL_TRANSITION
#define Z_SMF_CREATE_STATE(_entry, _run, _exit, _parent) \
{ \
	.entry   = _entry,  \
	.run     = _run,    \
	.exit    = _exit,   \
	.parent  = _parent, \
}
#define SMF_CREATE_STATE(_entry, _run, _exit, ...) Z_SMF_CREATE_STATE(_entry, _run, _exit, __VA_ARGS__)
#else
#define Z_SMF_CREATE_STATE(_entry, _run, _exit, _parent, _initial)\
{ \
	.entry   = _entry,  \
	.run     = _run,    \
	.exit    = _exit,   \
	.parent  = _parent, \
	.initial  = _initial, \
}
#define SMF_CREATE_STATE(_entry, _run, _exit, ...) Z_SMF_CREATE_STATE(_entry, _run, _exit, __VA_ARGS__)
#endif
#else
#define SMF_CREATE_STATE(_entry, _run, _exit, ...) \
{ \
	.entry   = _entry,  \
	.run     = _run,    \
	.exit    = _exit,   \
}
#endif
```

Could potentially be made into a more presentable state by using things like:
- `IS_ENABLED`
- `IS_EQ`
- `IF_ENABLED`
- `GET_ARG_N`

maass-hamburg commented:
> > The problem is, that `SMF_CREATE_STATE()` is a macro and not a function. Macros can't use `__VA_ARGS__` to get the single parameters.
> 
> i.e.:
> 
> ```c
> #ifdef CONFIG_SMF_ANCESTOR_SUPPORT
> #ifndef CONFIG_SMF_INITIAL_TRANSITION
> #define Z_SMF_CREATE_STATE(_entry, _run, _exit, _parent) \
> { \
> 	.entry   = _entry,  \
> 	.run     = _run,    \
> 	.exit    = _exit,   \
> 	.parent  = _parent, \
> }
> #define SMF_CREATE_STATE(_entry, _run, _exit, ...) Z_SMF_CREATE_STATE(_entry, _run, _exit, __VA_ARGS__)
> #else
> #define Z_SMF_CREATE_STATE(_entry, _run, _exit, _parent, _initial)\
> { \
> 	.entry   = _entry,  \
> 	.run     = _run,    \
> 	.exit    = _exit,   \
> 	.parent  = _parent, \
> 	.initial  = _initial, \
> }
> #define SMF_CREATE_STATE(_entry, _run, _exit, ...) Z_SMF_CREATE_STATE(_entry, _run, _exit, __VA_ARGS__)
> #endif
> #else
> #define SMF_CREATE_STATE(_entry, _run, _exit, ...) \
> { \
> 	.entry   = _entry,  \
> 	.run     = _run,    \
> 	.exit    = _exit,   \
> }
> #endif
> ```
> 
> Could potentially be made into a more presentable state by using things like:
> 
> * `IS_ENABLED`
> * `IS_EQ`
> * `IF_ENABLED`
> * `GET_ARG_N`

If the architecture WG is not approving the RFC, that could be an alternative. But having a constant number of arguments is still the more beautiful and better IMO, even if it breaks existing code (but easily fixable).

ycsin commented on include/zephyr/smf.h:
```diff
@@ -47,33 +31,14 @@
+	.entry = _entry,                                                       \
+	.run   = _run,                                                         \
+	.exit  = _exit,                                                        \
+	COND_CODE_1(CONFIG_SMF_ANCESTOR_SUPPORT, (.parent = _parent,), ())     \
```
could be further simplified
```suggestion
	IF_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT, (.parent = _parent,))     \
```

maass-hamburg commented:
The architecture WG has no objections:
https://github.com/zephyrproject-rtos/zephyr/issues/71252#issuecomment-2085585788
@sambhurst @keith-zephyr looking towards getting a approving review.

ycsin commented on doc/releases/release-notes-3.7.rst:
```diff
@@ -229,6 +229,10 @@ Libraries / Subsystems
 
+* State Machine Framework
+
+  * The :c:macro:`SMF_CREATE_STATE` macro now always takes 5 arguments.
```
I think we should also mention:
```rst
independent of the values of :kconfig:option:`CONFIG_SMF_ANCESTOR_SUPPORT` and
  :kconfig:option:`CONFIG_SMF_INITIAL_TRANSITION`
```

carlescufi commented:
@keith-zephyr please take another look
